bam_import.c

```diff
@@ -0,0 +1,475 @@
+    fprintf(fp, "  --i1 FILE    Index-1 from FILE\n");
+    fprintf(fp, "  --i2 FILE    Index-2 from FILE\n");
+    fprintf(fp, "  -i           Parse CASAVA identifier\n");
+    fprintf(fp, "  --barcode-tag TAG\n");
+    fprintf(fp, "               Tag to use with barcode sequences [BC]\n");
+    fprintf(fp, "  --quality-tag TAG\n");
+    fprintf(fp, "               Tag to use with barcode qualities [QT]\n");
+    fprintf(fp, "  -r STRING    Build up a complete @RG line\n");
+    fprintf(fp, "  -R STRING    Add a simple RG line of \"@RG\\tID:STRING\"\n");
+    fprintf(fp, "  -T TAGLIST   Parse tags in SAM format; list of '*' for all\n");
+    fprintf(fp, "  -o FILE      Output to FILE instead of stdout\n");
+    fprintf(fp, "  -u           Uncompressed output\n");
+    fprintf(fp, "  --order TAG  Store Nth record count in TAG\n");
+    fprintf(fp, "\n");
+    sam_global_opt_help(fp, "-.O.-@--");
+
+    fprintf(fp, "\nA single fastq file will be interpreted as -s, -0 or -1 depending on\n");
+    fprintf(fp, "file contents, and a pair of fastq files as \"-1 FILE1 -2 FILE2\".\n");
+
+    return exit_status;
+}
+
+// Order matters here as we want to read index elements before main
+// sequences so on reading the seqs we can emit a fully annotated record.
+enum fileno {
+    FQ_I1, FQ_I2, // index seqs for R1 and R2
+    FQ_R0,        // single file and unpaired data (singled-ended tech).
+    FQ_R1, FQ_R2, // separate read1 and read2 files
+    FQ_SINGLE,    // single file, but with read1 and/or read2 present.
+    FQ_END
+};
+
+typedef struct {
+    sam_global_args ga;
+    int no_pg;
+    char *fn[FQ_END], *fn_out;
+    int idx_both;      // add index to READ2 too, not just READ1
+    int casava;
+    char *barcode_seq;
+    char *barcode_qual;
+    char *aux;
+    char *rg;
+    char *rg_line;
+    char *order;
+    int compress_level;
+    htsThreadPool p;
+} opts_t;
+
+// Append a sequence and quality string from a BAM record to a BC:Z and
+// QT:Z style aux tag string.
+static int append_index(kstring_t *s, kstring_t *q, bam1_t *b) {
+    char *sp, *qp;
+    if (ks_resize(s, s->l + b->core.l_qseq+1 +1) < 0)
+        return -1;
+    if (ks_resize(q, q->l + b->core.l_qseq+1 +1) < 0)
+        return -1;
+
+    sp = s->s + s->l - (s->l > 0);
+    qp = q->s + q->l - (q->l > 0);
+
+    if (s->l)
+        *sp++ = '-';
+
+    if (q->l)
+        *qp++ = ' ';
+
+    int i;
+    uint8_t *seq = bam_get_seq(b);
+    uint8_t *qual = bam_get_qual(b);
+    for (i = 0; i < b->core.l_qseq; i++) {
+        *sp++ = seq_nt16_str[bam_seqi(seq, i)];
+        *qp++ = qual[i] + '!';
+    }
+    *sp++ = 0;
+    *qp++ = 0;
+
+    s->l = sp - s->s;
+    q->l = qp - q->s;
+
+    return 0;
+}
+
+static int import_fastq(int argc, char **argv, opts_t *opts) {
+    int i, n, ret = 0;
+    samFile *fp_in[FQ_END] = {NULL};
+    bam1_t *b = bam_init1();
+    int ids[FQ_END];
+    samFile *fp_out = NULL;
+    sam_hdr_t *hdr_out = NULL;
+    char index_str[1000] = {0};
```

Comment:
`index_str`, `read_str` and `rg_line` would be better implemented using `kstring_t`.